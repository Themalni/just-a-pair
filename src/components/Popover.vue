<template>
  <div class="custom-popover">
    <!-- activator goes here -->
    <div @click="openPopover" v-click-outside="closePopover" v-if="$slots.activator">
      <slot name="activator"></slot>
    </div>
    <!-- content goes here -->
    <div v-show="isOpen" v-if="$slots.content">
      <slot name="content"></slot>
    </div>
  </div>
</template>
<script>
import ClickOutside from "vue-click-outside";

export default {
  name: "VPopover",
  directives: {
    ClickOutside
  },
  data: () => ({
    isOpen: false
  }),
  methods: {
    openPopover() {
      this.isOpen = true;
    },
    closePopover() {
      this.isOpen = false;
    }
  }
};
</script>
<style lang="scss">
.custom-popover {
  position: relative;
}
.popover-content {
  position: absolute;
  top: 42px;
  left: 50%;
  transform: translateX(-50%);
  background-color: #fff;
  border-radius: 2px;
  box-shadow: 1px 1px 10px -2px $dark-grey;
}

.popover-activator {
  line-height: 20px;
  font-size: 14px;
  color: rgb(49, 49, 49);
  cursor: pointer;

  .popover-button-icon {
    vertical-align: middle;
  }
}

.popover-arrow {
  position: absolute;
  top: -12px;
  left: 50%;
  transform: translateX(-50%);
  display: block;
  width: 0;
  height: 0;
  border-width: 8px 12px;
  border-style: solid;
  border-color: transparent;
  border-bottom-color: #fff;
  z-index: 1;
}

.popover-item {
  display: block;
  min-width: 110px;
  padding: 1em;
  font-size: 1em;
  color: #666;
  position: relative;
  cursor: pointer;
  z-index: 2;

  &:nth-child(2) {
    border-top-left-radius: 2px;
    border-top-right-radius: 2px;
  }

  &:last-child {
    border-bottom-left-radius: 2px;
    border-bottom-right-radius: 2px;
  }

  &.popover-link {
    &:hover {
      text-decoration: none;
      background-color: $light-blue;
      color: #fff;
    }
  }
}

.poprover-icon {
  width: 15px;
  display: inline-block;
  text-align: center;
}

.popover-item-text {
  margin-left: 0.5em;
}

.popover-link {
  color: $dark-grey;
}
</style>
