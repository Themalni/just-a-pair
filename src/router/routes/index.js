export default [{
    path: '/',
    name: 'Home',
    component: () => import( /* webpackChunkName: "Home" */ '@/views/Home')
},
{
    path: '/product/:id',
    name: 'Product',
    component: () => import( /* webpackChunkName: "Product" */ '@/views/Product')
},
{
    path: '/user-dashboard',
    name: 'UserDashboard',
    component: () => import( /* webpackChunkName: "User Dashboard" */ '@/views/user/UserDashboard'),
    meta: {
        authRequired: true
    }
}, {
    path: '/admin-dashboard',
    name: 'AdminDashboard',
    component: () => import( /* webpackChunkName: "Admin Dashboard" */ '@/views/admin/AdminDashboard'),
    meta: {
        authRequired: true
    }
}
]