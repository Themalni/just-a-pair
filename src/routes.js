import ProductView from './components/ProductView.vue'
import HomePage from './components/HomePage.vue'


export default [
    {path: '/', component: HomePage},
    {path: '/product-view/:id', component: ProductView }
]