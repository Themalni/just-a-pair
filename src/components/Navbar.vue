<template>
  <div class="app-header">
    <!--Navbar-->
    <nav class="navbar navbar-dark info-color fixed-top scrolling-navbar">
      <!-- Navbar brand -->
      <router-link to="/" class="navbar-brand">
        <img :src="logo" />
      </router-link>
      <ul class="navbar-nav ml-auto navbar-icons">
        <li class="nav-item">
          <user-popover>
            <template slot="activator">
              <a class="nav-link font-weight-bolder text-large">
                User
                <md-icon>account_circle</md-icon>
              </a>
            </template>
            <template slot="content">
              <div class="popover-content">
                <span class="popover-arrow"></span>
                <a
                  v-if="!user"
                  class="popover-item popover-link font-weight-bolder"
                  href="#userLoginModal"
                  data-toggle="modal"
                  data-target="#userLoginModal"
                  @click="openRegisterTab()"
                >
                  <span class="popover-text">Register</span>
                </a>
                <a
                  v-if="!user"
                  class="popover-item popover-link font-weight-bolder"
                  href="#userLoginModal"
                  data-toggle="modal"
                  data-target="#userLoginModal"
                  @click="openLoginTab()"
                >
                  <span class="popover-text">Login</span>
                </a>
                <router-link
                  v-if="!!user"
                  class="popover-item popover-link font-weight-bolder"
                  to="/admin-dashboard"
                >Admin</router-link>
                <div v-if="!!user" class="popover-item popover-link">
                  <button class="button-reset font-weight-bolder" @click="logout()">Logout</button>
                </div>
              </div>
            </template>
          </user-popover>
        </li>
        <li class="nav-item">
          <a
            href="#shoppingCartModal"
            class="nav-link font-weight-bolder text-large"
            data-toggle="modal"
            data-target="#shoppingCartModal"
          >
            <span v-show="inCart.length > 0" class="cart-counter">{{ inCart.length }}</span>
            <span class="mr-1">Cart</span>
            <md-icon>shopping_cart</md-icon>
          </a>
        </li>
      </ul>
      <!-- Collapse button -->
      <button
        class="navbar-toggler"
        type="button"
        data-toggle="collapse"
        data-target="#basicExampleNav"
        aria-controls="basicExampleNav"
        aria-expanded="false"
        aria-label="Toggle navigation"
      >
        <span class="navbar-toggler-icon"></span>
      </button>
      <!-- Collapsible content -->
      <div class="collapse navbar-collapse" id="basicExampleNav">
        <!-- Links -->
        <ul class="navbar-nav mr-auto">
          <li class="nav-item active">
            <a class="nav-link" href="#">Home</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="#">
              Women
              <span class="sr-only">(current)</span>
            </a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="#">Men</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="#">Children</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="#">Sale</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="#">About Us</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="#">Your order</a>
          </li>
        </ul>
        <!-- Links -->
      </div>
      <!-- Collapsible content -->
    </nav>
    <!--/.Navbar-->
    <login />
  </div>
</template>

<script>
import Login from "@/components/user/Login";
import Popover from "@/components/Popover";
import { fb } from "@/firebase";
import { mapState } from "vuex";

export default {
  name: "Navbar",
  components: {
    Login,
    "user-popover": Popover
  },
  data() {
    return {
      logo: "../assets/logo.svg"
    };
  },
  computed: {
    ...mapState(["user"]),
    inCart() {
      return this.$store.getters.inCart;
    }
  },
  methods: {
    openRegisterTab() {
      $('#loginTab a[href="#register"]').tab("show");
    },
    openLoginTab() {
      $('#loginTab a[href="#login"]').tab("show");
    },
    async logout() {
      await this.$auth.logout();
      this.$router.push("/");
    }
  }
};
</script>

<style lang="scss" scoped>
.app-header {
  border-bottom: 1px solid $grey;
}
.navbar {
  &.navbar-dark {
    .navbar-nav {
      .nav-item {
        .nav-link {
          padding: 10px;
          position: relative;

          &:hover {
            text-decoration: none;
          }
        }
      }
    }
  }
  .navbar-brand {
    width: 30px;
  }
}
.navbar-nav {
  &.navbar-icons {
    flex-direction: row;
    flex-wrap: nowrap;

    .nav-item {
      display: inherit;
    }
  }
}
.info-color {
  background-color: $main-blue !important;
}
.cart-counter {
  width: 20px;
  height: 20px;
  border-radius: 50%;
  background-color: $orange-accent;
  text-align: center;
  font-weight: bold;
  font-size: 0.8em;
  color: #fff;
  position: absolute;
  top: 0;
  right: -0.2em;
}
</style>