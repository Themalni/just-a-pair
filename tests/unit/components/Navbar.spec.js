import { mount, createLocalVue } from '@vue/test-utils';
import Navbar from '@/components/Navbar';
import Popover from "@/components/Popover";
import mockState from '../utils/mockState';
import Vuex from 'vuex';
import salePrice from "@/filters.js";
import VueRouter from "vue-router";
import routes from "@/router/routes";
import $ from 'jquery';

global.$ = $;

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(VueRouter);
localVue.filter('salePrice', salePrice);

describe('Navbar', () => {
    let wrapper
    let store
    let state
    let router

    beforeEach(() => {
        router = new VueRouter({ routes })
        state = mockState.state
        store = new Vuex.Store({
            state: state
        })

        wrapper = mount(Navbar, {
            stubs: ['md-icon'],
            store,
            localVue,
            router
        });
    });

    afterEach(() => {
        wrapper.destroy();
    });

    it('renders correctrly', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('register button is visible when user is not logged in', async () => {
        state.user = null;
        const navPopover = wrapper.findComponent(Popover);
        const userButton = navPopover.findAll('a').at(0);

        await userButton.trigger('click');

        const registerButton = navPopover.find('.register-btn');

        expect(store.state.user).toBe(null);
        expect(registerButton.isVisible()).toBe(true);
    });

    it('login button is visible when user is not logged in', async () => {
        state.user = null;
        const navPopover = wrapper.findComponent(Popover);
        const userButton = navPopover.findAll('a').at(0);

        await userButton.trigger('click');

        const loginButton = navPopover.find('.login-btn');

        expect(store.state.user).toBe(null);
        expect(loginButton.isVisible()).toBe(true);
    });

    it('user profile button is visible when user is logged in', async () => {
        store.state.user = {
            displayName: 'John Smith',
            email: 'john_doe@gmail.com'
        }
        const navPopover = wrapper.findComponent(Popover);
        const userButton = navPopover.findAll('a').at(0);

        await userButton.trigger('click');

        const userProfileButton = await navPopover.find('.user-profile');

        expect(userProfileButton.isVisible()).toBe(true);
    });

    it('logout button is visible when user is logged in', async () => {
        store.state.user = {
            displayName: 'John Smith',
            email: 'john_doe@gmail.com'
        }
        const navPopover = wrapper.findComponent(Popover);
        const userButton = navPopover.findAll('a').at(0);

        await userButton.trigger('click');

        const logoutButton = await navPopover.find('.logout-btn');

        expect(logoutButton.isVisible()).toBe(true);
    });

    it('logs out user', async () => {
        store.state.user = {
            displayName: 'John Smith',
            email: 'john_doe@gmail.com'
        }
        const navPopover = wrapper.findComponent(Popover);
        const userButton = navPopover.findAll('a').at(0);
        const logoutMock = jest.spyOn(wrapper.vm, 'logout');
        const firebaseSignOutMock = jest.spyOn(wrapper.vm, 'firebaseSignOut');

        await userButton.trigger('click');

        const logoutButton = navPopover.find('.logout-btn');

        await logoutButton.trigger("click");

        expect(logoutMock).toHaveBeenCalled();
        expect(firebaseSignOutMock).toHaveBeenCalled();
    })
})