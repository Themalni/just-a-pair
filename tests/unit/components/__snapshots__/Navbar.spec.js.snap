// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctrly 1`] = `
<div class="app-header">
  <nav class="navbar navbar-dark info-color fixed-top scrolling-navbar"><a href="#/" aria-current="page" class="navbar-brand router-link-exact-active router-link-active"><img src="../assets/logo.svg"></a>
    <ul class="navbar-nav ml-auto navbar-icons">
      <li class="nav-item">
        <div class="custom-popover">
          <div><a class="nav-link font-weight-bolder text-large">
              User
              <md-icon-stub>account_circle</md-icon-stub>
            </a></div>
          <div style="display: none;">
            <div class="popover-content"><span class="popover-arrow"></span>
              <!---->
              <!----> <a href="#/user-dashboard" class="user-profile popover-item popover-link font-weight-bolder">User profile</a>
              <div class="popover-item popover-link"><button class="logout-btn button-reset font-weight-bolder">Logout</button></div>
            </div>
          </div>
        </div>
      </li>
      <li class="nav-item"><a href="#shoppingCartModal" data-toggle="modal" data-target="#shoppingCartModal" class="nav-link font-weight-bolder text-large"><span class="cart-counter">1</span> <span class="mr-1">Cart</span>
          <md-icon-stub>shopping_cart</md-icon-stub>
        </a></li>
    </ul> <button type="button" data-toggle="collapse" data-target="#basicExampleNav" aria-controls="basicExampleNav" aria-expanded="false" aria-label="Toggle navigation" class="navbar-toggler"><span class="navbar-toggler-icon"></span></button>
    <div id="basicExampleNav" class="collapse navbar-collapse">
      <ul class="navbar-nav mr-auto">
        <li class="nav-item active"><a href="#" class="nav-link">Home</a></li>
        <li class="nav-item"><a href="#" class="nav-link">
            Women
            <span class="sr-only">(current)</span></a></li>
        <li class="nav-item"><a href="#" class="nav-link">Men</a></li>
        <li class="nav-item"><a href="#" class="nav-link">Children</a></li>
        <li class="nav-item"><a href="#" class="nav-link">Sale</a></li>
        <li class="nav-item"><a href="#" class="nav-link">About Us</a></li>
        <li class="nav-item"><a href="#" class="nav-link">Your order</a></li>
      </ul>
    </div>
  </nav>
  <div id="userLoginModal" tabindex="-1" role="dialog" aria-hidden="true" class="modal fade">
    <div role="document" class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header text-center">
          <ul id="loginTab" role="tablist" class="nav nav-tabs">
            <li class="nav-item"><a id="login-tab" data-toggle="tab" href="#login" role="tab" aria-controls="login" aria-selected="true" class="nav-link active mr-1">Login</a></li>
            <li class="nav-item"><a id="register-tab" data-toggle="tab" href="#register" role="tab" aria-controls="register" aria-selected="false" class="nav-link ml-1">Register</a></li>
          </ul>
        </div>
        <div class="modal-body mx-3">
          <div id="myTabContent" class="tab-content">
            <div id="login" role="tabpanel" aria-labelledby="login-tab" class="tab-pane fade show active">
              <div class="md-form mb-2 mt-3">
                <md-icon-stub class="prefix grey-text">email</md-icon-stub> <input type="email" id="login-email" class="form-control validate mb-2"> <label data-error="wrong" data-success="right" for="login-email">Your email</label>
              </div>
              <div class="md-form mb-2">
                <md-icon-stub class="prefix grey-text">lock</md-icon-stub> <input type="password" id="login-pass" class="form-control validate mb-2"> <label data-error="wrong" data-success="right" for="login-pass">Your password</label>
              </div> <button class="login-btn btn btn-primary btn-block mb-2 mt-4">Login</button>
            </div>
            <div id="register" role="tabpanel" aria-labelledby="register-tab" class="tab-pane fade">
              <div class="md-form pb-2 mt-2">
                <md-icon-stub class="prefix grey-text">account_circle</md-icon-stub> <input type="text" id="register-name" class="form-control validate mb-2"> <label data-error="wrong" data-success="right" for="register-name">Your name</label>
              </div>
              <div class="md-form pb-2 mt-2">
                <md-icon-stub class="prefix grey-text">email</md-icon-stub> <input type="email" id="register-email" class="form-control validate mb-2 mt-4"> <label data-error="wrong" data-success="right" for="register-email">Your email</label>
              </div>
              <div class="md-form pb-2 mt-2">
                <md-icon-stub class="prefix grey-text">lock</md-icon-stub> <input type="password" id="register-pass" class="form-control validate mb-2"> <label data-error="wrong" data-success="right" for="register-pass">Your password</label>
              </div> <button class="register-btn btn btn-default btn-block mb-2 mt-4">Register</button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
